packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,116 @@
+import { PhysicalNameGenerator } from "./physical-name-generator";
+import { IResource } from './resource';
+import { Token } from "./token";
+
+/**
+ * Options allowing customizing the automatically generated physical name of the resource.
+ * Used in the {@link PhysicalName#auto} method.
+ */
+export interface AutoPhysicalNameOptions {
+  /**
+   * Can the framework generate and set a physical name for the resource automatically
+   * if it's used in a cross-environment fashion.
+   *
+   * @default false
+   */
+  readonly generateIfCrossEnv?: boolean;
+}
+
+/**
+ * Represents the human-readable, physical name of a resource.
+ * These can be known at code runtime,
+ * in the case of fixed, customer-provided names,
+ * at synthesis time, in case of automatically assigned names by the framework,
+ * or only at deploy time, assigned by CloudFormation -
+ * which is the default.
+ *
+ * @see #auto()
+ * @see #of()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically generated.
+   * The details of how and when that generation will occur can be customized by passing an instance
+   * of {@link AutoPhysicalNameOptions}.
+   * By default, the physical name of the resource will be generated by CloudFormation,
+   * at deploy time.
+   */
+  public static auto(options: AutoPhysicalNameOptions = {}): PhysicalName {
+    return new LateBoundPhysicalName(options);
+  }
+
+  /**
+   * A fixed physical name (one that is known statically, at synthesis time).
+   *
+   * @param name the name to assign
+   */
+  public static of(name: string): PhysicalName {
+    return new FixedPhysicalName(name);
+  }
+
+  protected constructor() {
+  }
+
+  /**
+   * Returns the physical name, as a String.
+   * If the customer assigned a physical name to the resource, it will be returned, unchanged;
+   * can be `undefined`, if a physical name was not provided,
+   * or an opaque Token, if a generated name should be assigned lazily at synthesis time.
+   */
+  public abstract get value(): string | undefined;
+
+  /** @internal */
+  public abstract _resolveForCrossEnv(resource: IResource): void;
+}
+
+// implementations are private to this module,
+// we only surface them through static factory methods
+
+class FixedPhysicalName extends PhysicalName {
+  private readonly name: string;
+
+  constructor(name: string) {
+    super();
+
+    this.name = name;
+  }
+
+  public get value(): string | undefined {
+    return this.name;
+  }
+
+  /** @internal */
+  public _resolveForCrossEnv(_resource: IResource): void {
+    // nothing to do - the name will always be the same
+  }
+}
+
+class LateBoundPhysicalName extends PhysicalName {
+  private readonly tokenString?: string;
+  private name?: string;
+
+  constructor(options: AutoPhysicalNameOptions) {
+    super();
+
+    this.tokenString = options.generateIfCrossEnv
```

Comment:
why not just assign directly to `this.value`. What's the rational for the indirection?

Reply:
Changed.